worker/worker-news-processor/index.js

```diff
@@ -558,116 +558,269 @@ async function processArticle(env, article, config) {
+      idIndex = idIndex.slice(0, maxArticles);
+    }
     
-    let processedCount = 0;
+    await kv.put(
+      config.KV_KEY_IDS,
+      JSON.stringify(idIndex),
+      {
+        expirationTtl: config.ID_INDEX_TTL
+      }
+    );
+  }
+  
+  // Step 7: Update checkpoint based on success/failure
+  if (processingFailed) {
+    // Move article to try-later list
+    checkpoint.tryLater = (checkpoint.tryLater || []).filter(item => item.id !== nextArticleId);
+    checkpoint.tryLater.push({
+      id: nextArticleId,
+      article: updatedArticle,
+      failedAt: Date.now(),
+      reason: failureReason
+    });
     
-    // Process articles and collect updates
-    const updatePromises = [];
+    if (!checkpoint.processedIds.includes(nextArticleId)) {
+      checkpoint.processedIds.push(nextArticleId);
+    }
     
-    for (const item of articlesToProcess) {
-      const { id, article } = item;
-      console.log(`\nProcessing article ${processedCount + 1}/${articlesToProcess.length}: "${article.title?.substring(0, 50)}..."`);
-      
-      try {
-        // Process the article (one phase per run)
-        const updatedArticle = await processArticle(env, article, config);
-        
-        // Queue the write operation
-        const articleKey = `article:${id}`;
-        updatePromises.push(
-          env.CRYPTO_NEWS_CACHE.put(articleKey, JSON.stringify(updatedArticle), {
-            expirationTtl: config.ID_INDEX_TTL || 60 * 60 * 24 * 30
-          })
-        );
-        
-        processedCount++;
-      } catch (error) {
-        console.error(`  ✗ Error processing article:`, error);
-        // Continue with next article
+    checkpoint.currentArticleId = null;
+    checkpoint.currentArticle = null;
+  } else {
+    // Check if article is fully complete
+    const isComplete = 
+      !(updatedArticle.needsSentiment ?? true) && 
+      !(updatedArticle.needsSummary ?? true) && 
+      !(updatedArticle.contentTimeout && updatedArticle.contentTimeout < config.MAX_CONTENT_FETCH_ATTEMPTS);
+    
+    if (isComplete) {
+      if (!checkpoint.processedIds.includes(nextArticleId)) {
+        checkpoint.processedIds.push(nextArticleId);
       }
+      checkpoint.currentArticleId = null;
+      checkpoint.currentArticle = null;
+      checkpoint.tryLater = (checkpoint.tryLater || []).filter(item => item.id !== nextArticleId);
+    } else {
+      // Keep in checkpoint for next run
+      checkpoint.currentArticle = updatedArticle;
     }
-    
-    // Batch write: Execute all updates in parallel
-    if (updatePromises.length > 0) {
-      await Promise.all(updatePromises);
-      console.log(`\n✓ Batch update: ${processedCount} articles written to KV`);
+  }
+  
+  checkpoint.lastUpdate = Date.now();
+  
+  await kv.put(
+    config.KV_KEY_CHECKPOINT,
+    JSON.stringify(checkpoint),
+    {
+      expirationTtl: config.ID_INDEX_TTL
     }
+  );
+  
+  return { processed: true, articleId: nextArticleId, article: updatedArticle };
```

Comment:
we could return the checkpoint as well so that the calling function can generate better logs without using another read from the KV